File: trading_agents/api/routes/agent_routes.py
"""
API routes for agent management.
"""
from fastapi import APIRouter, HTTPException, Depends, Query, WebSocket, WebSocketDisconnect
from typing import List, Dict, Any, Optional
import asyncio
import json
import logging
from datetime import datetime

from core.agent_registry import AgentRegistry
from database import get_agent_repository, get_position_repository, get_market_repository
from api.models.agent_models import (
    AgentConfig,
    AgentConfigUpdate,
    AgentAction,
    AgentStatus,
    AgentResponse,
    AgentListResponse,
    ActionResponse
)

# Set up logging
logger = logging.getLogger(__name__)

# Create router
router = APIRouter(prefix="/agents", tags=["agents"])

# WebSocket connections
websocket_connections: Dict[str, List[WebSocket]] = {}

# Get agent registry
async def get_agent_registry() -> AgentRegistry:
    """
    Get the agent registry.
    """
    # Get the singleton instance
    return AgentRegistry.get_instance()

@router.post("/", response_model=AgentResponse)
async def create_agent(
    agent_config: AgentConfig,
    registry: AgentRegistry = Depends(get_agent_registry)
):
    """
    Create a new agent.
    """
    try:
        # Generate agent ID
        agent_id = f"{agent_config.agent_type}_{datetime.now().strftime('%Y%m%d%H%M%S')}"

        # Register agent
        agent = await registry.register_agent(
            agent_config.agent_type,
            agent_id,
            {
                "name": agent_config.name,
                **agent_config.config
            }
        )

        # Get agent status
        status = await agent.get_status()

        return {
            "agent_id": agent_id,
            "name": agent_config.name,
            "agent_type": agent_config.agent_type,
            "status": status["status"],
            "created_at": datetime.now(),
            "updated_at": datetime.now(),
            "config": status["config"],
            "metrics": status["metrics"]
        }
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.error(f"Error creating agent: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Internal server error: {str(e)}")

@router.get("/", response_model=AgentListResponse)
async def get_agents(
    agent_type: Optional[str] = None,
    status: Optional[str] = None,
    limit: int = Query(100, ge=1, le=1000),
    offset: int = Query(0, ge=0),
    registry: AgentRegistry = Depends(get_agent_registry)
):
    """
    Get all agents with optional filtering and pagination.

    Args:
        agent_type: Optional filter by agent type
        status: Optional filter by agent status
        limit: Maximum number of agents to return
        offset: Pagination offset
        registry: Agent registry
    """
    try:
        # Use the new get_agents method that uses the database
        result = await registry.get_agents(
            agent_type=agent_type,
            status=status,
            limit=limit,
            offset=offset
        )

        # Return the result directly
        return {
            "agents": result["agents"],
            "total": result["total"]
        }
    except Exception as e:
        logger.error(f"Error getting agents: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Internal server error: {str(e)}")

@router.get("/{agent_id}", response_model=AgentResponse)
async def get_agent(
    agent_id: str,
    registry: AgentRegistry = Depends(get_agent_registry)
):
    """
    Get an agent by ID.
    """
    try:
        agent = await registry.get_agent(agent_id)

        if not agent:
            raise HTTPException(status_code=404, detail=f"Agent {agent_id} not found")

        status = await agent.get_status()

        return {
            "agent_id": agent.agent_id,
            "name": status["config"].get("name", "Unnamed Agent"),
            "agent_type": agent.__class__.__name__.replace("Agent", "").lower(),
            "status": status["status"],
            "created_at": datetime.now(),  # In a real app, this would be stored in the agent
            "updated_at": datetime.now(),  # In a real app, this would be stored in the agent
            "config": status["config"],
            "metrics": status["metrics"],
            "logs": agent.logger.get_logs(limit=100)
        }
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error getting agent {agent_id}: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Internal server error: {str(e)}")

@router.delete("/{agent_id}", response_model=ActionResponse)
async def delete_agent(
    agent_id: str,
    registry: AgentRegistry = Depends(get_agent_registry)
):
    """
    Delete an agent.
    """
    try:
        await registry.unregister_agent(agent_id)

        return {
            "success": True,
            "message": f"Agent {agent_id} deleted successfully"
        }
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        logger.error(f"Error deleting agent {agent_id}: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Internal server error: {str(e)}")

@router.put("/{agent_id}", response_model=AgentResponse)
async def update_agent(
    agent_id: str,
    agent_config: AgentConfigUpdate,
    registry: AgentRegistry = Depends(get_agent_registry)
):
    """
    Update an agent.
    """
    try:
        agent = await registry.get_agent(agent_id)

        if not agent:
            raise HTTPException(status_code=404, detail=f"Agent {agent_id} not found")

        # Get current status
        current_status = await agent.get_status()
        current_config = current_status["config"]

        # Update config
        new_config = {**current_config}

        if agent_config.name:
            new_config["name"] = agent_config.name

        if agent_config.config:
            for key, value in agent_config.config.items():
                new_config[key] = value

        # Update agent
        await agent.update_config(new_config)

        # Get updated status
        status = await agent.get_status()

        return {
            "agent_id": agent.agent_id,
            "name": status["config"].get("name", "Unnamed Agent"),
            "agent_type": agent.__class__.__name__.replace("Agent", "").lower(),
            "status": status["status"],
            "created_at": datetime.now(),  # In a real app, this would be stored in the agent
            "updated_at": datetime.now(),  # In a real app, this would be stored in the agent
            "config": status["config"],
            "metrics": status["metrics"]
        }
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error updating agent {agent_id}: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Internal server error: {str(e)}")

@router.post("/{agent_id}/start", response_model=AgentResponse)
async def start_agent(
    agent_id: str,
    registry: AgentRegistry = Depends(get_agent_registry)
):
    """
    Start an agent.
    """
    try:
        # Use the updated start_agent method from the registry
        result = await registry.start_agent(agent_id)

        if not result["success"]:
            raise HTTPException(status_code=500, detail=result.get("error", "Unknown error"))

        # Get the agent to return the full response
        agent = await registry.get_agent(agent_id)

        if not agent:
            raise HTTPException(status_code=404, detail=f"Agent {agent_id} not found")

        # Get updated status
        status = await agent.get_status()

        return {
            "agent_id": agent.agent_id,
            "name": status["config"].get("name", "Unnamed Agent"),
            "agent_type": agent.__class__.__name__.replace("Agent", "").lower(),
            "status": status["status"],
            "created_at": datetime.now(),  # In a real app, this would be stored in the agent
            "updated_at": datetime.now(),  # In a real app, this would be stored in the agent
            "config": status["config"],
            "metrics": status["metrics"]
        }
    except HTTPException:
        raise
    except ValueError as e:
        if "not found" in str(e):
            raise HTTPException(status_code=404, detail=str(e))
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.error(f"Error starting agent {agent_id}: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Internal server error: {str(e)}")

@router.post("/{agent_id}/stop", response_model=AgentResponse)
async def stop_agent(
    agent_id: str,
    registry: AgentRegistry = Depends(get_agent_registry)
):
    """
    Stop an agent.
    """
    try:
        # Use the updated stop_agent method from the registry
        result = await registry.stop_agent(agent_id)

        if not result["success"]:
            raise HTTPException(status_code=500, detail=result.get("error", "Unknown error"))

        # Get the agent to return the full response
        agent = await registry.get_agent(agent_id)

        if not agent:
            raise HTTPException(status_code=404, detail=f"Agent {agent_id} not found")

        # Get updated status
        status = await agent.get_status()

        return {
            "agent_id": agent.agent_id,
            "name": status["config"].get("name", "Unnamed Agent"),
            "agent_type": agent.__class__.__name__.replace("Agent", "").lower(),
            "status": status["status"],
            "created_at": datetime.now(),  # In a real app, this would be stored in the agent
            "updated_at": datetime.now(),  # In a real app, this would be stored in the agent
            "config": status["config"],
            "metrics": status["metrics"]
        }
    except HTTPException:
        raise
    except ValueError as e:
        if "not found" in str(e):
            raise HTTPException(status_code=404, detail=str(e))
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.error(f"Error stopping agent {agent_id}: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Internal server error: {str(e)}")

@router.get("/{agent_id}/status", response_model=AgentStatus)
async def get_agent_status(
    agent_id: str,
    registry: AgentRegistry = Depends(get_agent_registry)
):
    """
    Get agent status.
    """
    try:
        agent = await registry.get_agent(agent_id)

        if not agent:
            raise HTTPException(status_code=404, detail=f"Agent {agent_id} not found")

        status = await agent.get_status()

        return status
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error getting agent status {agent_id}: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Internal server error: {str(e)}")

@router.post("/{agent_id}/execute", response_model=ActionResponse)
async def execute_agent_action(
    agent_id: str,
    action: AgentAction,
    registry: AgentRegistry = Depends(get_agent_registry)
):
    """
    Execute an action on an agent.
    """
    try:
        agent = await registry.get_agent(agent_id)

        if not agent:
            raise HTTPException(status_code=404, detail=f"Agent {agent_id} not found")

        result = await agent.execute_action({
            "type": action.type,
            **action.parameters
        })

        return result
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error executing action on agent {agent_id}: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Internal server error: {str(e)}")

@router.get("/{agent_id}/logs")
async def get_agent_logs(
    agent_id: str,
    limit: int = Query(100, ge=1, le=1000),
    level: Optional[str] = None,
    offset: int = Query(0, ge=0),
    registry: AgentRegistry = Depends(get_agent_registry)
):
    """
    Get agent logs.

    Args:
        agent_id: Agent ID
        limit: Maximum number of logs to return
        level: Optional filter by log level
        offset: Pagination offset
        registry: Agent registry
    """
    try:
        # Check if agent exists
        agent = await registry.get_agent(agent_id)
        if not agent:
            raise HTTPException(status_code=404, detail=f"Agent {agent_id} not found")

        # Get agent repository
        agent_repo = get_agent_repository()

        # Get logs from database
        logs = await agent_repo.get_agent_logs(
            agent_id=agent_id,
            level=level,
            limit=limit,
            offset=offset
        )

        return logs
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error getting agent logs {agent_id}: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Internal server error: {str(e)}")

@router.websocket("/{agent_id}/ws")
async def websocket_endpoint(
    websocket: WebSocket,
    agent_id: str,
    registry: AgentRegistry = Depends(get_agent_registry)
):
    """
    WebSocket endpoint for real-time agent updates.
    """
    await websocket.accept()

    # Check if agent exists
    agent = await registry.get_agent(agent_id)
    if not agent:
        await websocket.send_text(json.dumps({
            "error": f"Agent {agent_id} not found"
        }))
        await websocket.close()
        return

    # Add connection to list
    if agent_id not in websocket_connections:
        websocket_connections[agent_id] = []
    websocket_connections[agent_id].append(websocket)

    try:
        while True:
            # Send agent status every 5 seconds
            status = await agent.get_status()
            await websocket.send_text(json.dumps(status))
            await asyncio.sleep(5)
    except WebSocketDisconnect:
        # Remove connection from list
        websocket_connections[agent_id].remove(websocket)
        if not websocket_connections[agent_id]:
            del websocket_connections[agent_id]
    except Exception as e:
        logger.error(f"WebSocket error for agent {agent_id}: {str(e)}")
        await websocket.close()

# Function to broadcast updates to all connected clients
async def broadcast_update(agent_id: str, data: Dict[str, Any]):
    """
    Broadcast an update to all connected clients for an agent.
    """
    if agent_id in websocket_connections:
        for connection in websocket_connections[agent_id]:
            try:
                await connection.send_text(json.dumps(data))
            except Exception as e:
                logger.error(f"Error broadcasting to WebSocket: {str(e)}")

File: backend/src/routes/agentRoutes.ts
/**
 * Routes for managing trading agents.
 */
import express from 'express';
import { AgentService } from '../services/agentService';

const router = express.Router();

/**
 * Get all agents.
 */
router.get('/', async (req, res) => {
  try {
    const agents = await AgentService.getAllAgents();
    res.json(agents);
  } catch (error) {
    console.error('Error getting agents:', error);
    res.status(500).json({
      error: 'Internal Server Error',
      message: error instanceof Error ? error.message : 'Unknown error'
    });
  }
});

/**
 * Get agent by ID.
 */
router.get('/:id', async (req, res) => {
  try {
    const { id } = req.params;
    const agent = await AgentService.getAgent(id);
    res.json(agent);
  } catch (error) {
    console.error(`Error getting agent ${req.params.id}:`, error);
    
    // Check if it's a 404 error
    if (error instanceof Error && error.message.includes('not found')) {
      return res.status(404).json({
        error: 'Not Found',
        message: `Agent with ID ${req.params.id} not found`
      });
    }
    
    res.status(500).json({
      error: 'Internal Server Error',
      message: error instanceof Error ? error.message : 'Unknown error'
    });
  }
});

/**
 * Create a new agent.
 */
router.post('/', async (req, res) => {
  try {
    const { name, type, config } = req.body;
    
    // Validate request body
    if (!name || !type) {
      return res.status(400).json({
        error: 'Bad Request',
        message: 'Name and type are required'
      });
    }
    
    const agent = await AgentService.createAgent({
      name,
      type,
      config: config || {}
    });
    
    res.status(201).json(agent);
  } catch (error) {
    console.error('Error creating agent:', error);
    res.status(500).json({
      error: 'Internal Server Error',
      message: error instanceof Error ? error.message : 'Unknown error'
    });
  }
});

/**
 * Update an agent.
 */
router.put('/:id', async (req, res) => {
  try {
    const { id } = req.params;
    const { name, config } = req.body;
    
    // Validate request body
    if (!name && !config) {
      return res.status(400).json({
        error: 'Bad Request',
        message: 'At least one of name or config must be provided'
      });
    }
    
    const agent = await AgentService.updateAgent(id, {
      name,
      config
    });
    
    res.json(agent);
  } catch (error) {
    console.error(`Error updating agent ${req.params.id}:`, error);
    
    // Check if it's a 404 error
    if (error instanceof Error && error.message.includes('not found')) {
      return res.status(404).json({
        error: 'Not Found',
        message: `Agent with ID ${req.params.id} not found`
      });
    }
    
    res.status(500).json({
      error: 'Internal Server Error',
      message: error instanceof Error ? error.message : 'Unknown error'
    });
  }
});

/**
 * Delete an agent.
 */
router.delete('/:id', async (req, res) => {
  try {
    const { id } = req.params;
    const success = await AgentService.deleteAgent(id);
    
    if (success) {
      res.status(204).send();
    } else {
      res.status(404).json({
        error: 'Not Found',
        message: `Agent with ID ${id} not found`
      });
    }
  } catch (error) {
    console.error(`Error deleting agent ${req.params.id}:`, error);
    res.status(500).json({
      error: 'Internal Server Error',
      message: error instanceof Error ? error.message : 'Unknown error'
    });
  }
});

/**
 * Start an agent.
 */
router.post('/:id/start', async (req, res) => {
  try {
    const { id } = req.params;
    const agent = await AgentService.startAgent(id);
    res.json(agent);
  } catch (error) {
    console.error(`Error starting agent ${req.params.id}:`, error);
    
    // Check if it's a 404 error
    if (error instanceof Error && error.message.includes('not found')) {
      return res.status(404).json({
        error: 'Not Found',
        message: `Agent with ID ${req.params.id} not found`
      });
    }
    
    res.status(500).json({
      error: 'Internal Server Error',
      message: error instanceof Error ? error.message : 'Unknown error'
    });
  }
});

/**
 * Stop an agent.
 */
router.post('/:id/stop', async (req, res) => {
  try {
    const { id } = req.params;
    const agent = await AgentService.stopAgent(id);
    res.json(agent);
  } catch (error) {
    console.error(`Error stopping agent ${req.params.id}:`, error);
    
    // Check if it's a 404 error
    if (error instanceof Error && error.message.includes('not found')) {
      return res.status(404).json({
        error: 'Not Found',
        message: `Agent with ID ${req.params.id} not found`
      });
    }
    
    res.status(500).json({
      error: 'Internal Server Error',
      message: error instanceof Error ? error.message : 'Unknown error'
    });
  }
});

/**
 * Get agent status.
 */
router.get('/:id/status', async (req, res) => {
  try {
    const { id } = req.params;
    const status = await AgentService.getAgentStatus(id);
    res.json(status);
  } catch (error) {
    console.error(`Error getting agent status ${req.params.id}:`, error);
    
    // Check if it's a 404 error
    if (error instanceof Error && error.message.includes('not found')) {
      return res.status(404).json({
        error: 'Not Found',
        message: `Agent with ID ${req.params.id} not found`
      });
    }
    
    res.status(500).json({
      error: 'Internal Server Error',
      message: error instanceof Error ? error.message : 'Unknown error'
    });
  }
});

/**
 * Execute an action on an agent.
 */
router.post('/:id/execute', async (req, res) => {
  try {
    const { id } = req.params;
    const { type, parameters } = req.body;
    
    // Validate request body
    if (!type) {
      return res.status(400).json({
        error: 'Bad Request',
        message: 'Action type is required'
      });
    }
    
    const result = await AgentService.executeAgentAction(id, type, parameters || {});
    res.json(result);
  } catch (error) {
    console.error(`Error executing action on agent ${req.params.id}:`, error);
    
    // Check if it's a 404 error
    if (error instanceof Error && error.message.includes('not found')) {
      return res.status(404).json({
        error: 'Not Found',
        message: `Agent with ID ${req.params.id} not found`
      });
    }
    
    res.status(500).json({
      error: 'Internal Server Error',
      message: error instanceof Error ? error.message : 'Unknown error'
    });
  }
});

/**
 * Get agent logs.
 */
router.get('/:id/logs', async (req, res) => {
  try {
    const { id } = req.params;
    const { limit, level } = req.query;
    
    const logs = await AgentService.getAgentLogs(
      id,
      limit ? parseInt(limit as string, 10) : undefined,
      level as string | undefined
    );
    
    res.json(logs);
  } catch (error) {
    console.error(`Error getting agent logs ${req.params.id}:`, error);
    
    // Check if it's a 404 error
    if (error instanceof Error && error.message.includes('not found')) {
      return res.status(404).json({
        error: 'Not Found',
        message: `Agent with ID ${req.params.id} not found`
      });
    }
    
    res.status(500).json({
      error: 'Internal Server Error',
      message: error instanceof Error ? error.message : 'Unknown error'
    });
  }
});

export default router;
